Gifski/ConversionViewController.swift

```diff
@@ -0,0 +1,124 @@
+	private var isRunning = false
+
+	convenience init(conversion: Gifski.Conversion) {
+		self.init()
+
+		self.conversion = conversion
+	}
+
+	override func loadView() {
+		let wrapper = NSView(frame: CGRect(origin: .zero, size: Constants.defaultWindowSize))
+		wrapper.translatesAutoresizingMaskIntoConstraints = false
+		wrapper.addSubview(circularProgress)
+		wrapper.addSubview(timeRemainingLabel)
+
+		circularProgress.constrain(to: CGSize(widthHeight: circularProgress.frame.width))
+		circularProgress.center(inView: wrapper)
+		NSLayoutConstraint.activate([
+			timeRemainingLabel.topAnchor.constraint(greaterThanOrEqualTo: circularProgress.bottomAnchor),
+			timeRemainingLabel.centerXAnchor.constraint(equalTo: circularProgress.centerXAnchor),
+			timeRemainingLabel.bottomAnchor.constraint(equalTo: wrapper.bottomAnchor, constant: -16.0)
+		])
+
+		view = wrapper
+	}
+
+	override func viewDidLoad() {
+		super.viewDidLoad()
+
+		start(conversion: conversion)
+	}
+
+	/// Gets called when the Esc key is pressed.
+	/// Reference: https://stackoverflow.com/a/42440020
+	@objc
+	func cancel(_ sender: Any?) {
+		cancelConversion()
+	}
+
+	private func start(conversion: Gifski.Conversion) {
+		guard !isRunning else {
+			return
+		}
+
+		isRunning = true
+
+		progress = Progress(totalUnitCount: 1)
+		progress?.publish()
+
+		circularProgress.progressInstance = progress
+		DockProgress.progressInstance = progress
+		timeRemainingEstimator.progress = progress
+		timeRemainingEstimator.start()
+
+		progress?.performAsCurrent(withPendingUnitCount: 1) {
+			Gifski.run(conversion) { result in
+				do {
+					let gifUrl = self.generateTempGifUrl(for: conversion.video)
+					try result.get().write(to: gifUrl, options: .atomic)
+					try? gifUrl.setMetadata(key: .itemCreator, value: "\(App.name) \(App.version)")
+					defaults[.successfulConversionsCount] += 1
+
+					self.didComplete(conversion: conversion, gifUrl: gifUrl)
+				} catch Gifski.Error.cancelled {
+					self.cancelConversion()
+				} catch {
+					self.presentError(error, modalFor: self.view.window)
+					self.cancelConversion()
+				}
+			}
+		}
+	}
+
+	private func generateTempGifUrl(for videoUrl: URL) -> URL {
+		let tempDirectory = FileManager.default.temporaryDirectory
+		let tempName = "\(videoUrl.filenameWithoutExtension).\(FileType.gif.fileExtension)"
+
+		return tempDirectory.appendingPathComponent(tempName, isDirectory: false)
+	}
+
+	private func cancelConversion() {
+		let videoDropController = VideoDropViewController()
+		stopConversion {
+			self.push(viewController: videoDropController)
+		}
+	}
+
+	private func didComplete(conversion: Gifski.Conversion, gifUrl: URL) {
+		let conversionCompleted = ConversionCompletedViewController(conversion: conversion, gifUrl: gifUrl)
+		stopConversion {
+			self.push(viewController: conversionCompleted)
+		}
+	}
+
+	private func stopConversion(_ completion: (() -> Void)? = nil) {
+		isRunning = false
+		progress?.cancel()
+		progress?.unpublish()
```

Comment:
To fix the crash when canceling, you need to do this in a next tick. The reason is that after `.cancel()`, the libgifski side needs to cancel too. Otherwise they will continue sending their last update, but `progress` will no longer exist and it will crash.
```suggestion

		DispatchQueue.main.async {
			self.progress?.unpublish()
		}

```

Reply:
Hmm, tried that and still got the crash sometimes 🤔

Reply:
What kind of crash though? Maybe it's a different one. Can you share?

Reply:
<img width="1220" alt="Zrzut ekranu 2019-07-8 o 14 57 15" src="https://user-images.githubusercontent.com/5232779/60811976-c2f96e80-a190-11e9-8dee-945ae08b1b1a.png">